### ISSUE TITLE
Exception: Serialization of 'SimpleXMLElement' is not allowed

### ISSUE THREAD

**Issue by followmeongitlab** [farmOS/farmOS#896]
**Describe the bug**
When trying to import a KML or KMZ file, I receive the error:

```
Exception: Serialization of 'SimpleXMLElement' is not allowed in serialize() (line 14 of /opt/drupal/web/core/lib/Drupal/Component/Serialization/PhpSerialize.php).
```

**To Reproduce**
1. Go to `Setup > Import > KML Import`
2. Click on `Browse` and select the KML file
3. Select `Property` from Default Land Type drop-down
4. Click `Parse`
5. See error

**Expected behavior**
I am expecting this to successfully import a KML file to use as a new land asset.

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04
 - Browser: FireFox
 - Version 132.0.1

**Additional context**
The KML file in question was an export of a single parcel from VT's parcel maps.
https://geodata.vermont.gov/datasets/09cf47e1cf82465e99164762a04f3ce6_0/explore

This is a new install (I am a fresh FarmOS user) using Docker/Docker Compose on a local server. Only modifications I've made to settings so far have been to enable file import.

 The docker-compose.yml:
 ```
 version: "3"
services:
  db:
    image: postgres:13
    restart: always
    volumes:
      - '/data/farmos/db:/var/lib/postgresql/data'
    ports:
      - '5433:5432'
    environment:
      POSTGRES_USER: farm
      POSTGRES_PASSWORD: farm
      POSTGRES_DB: farm
      POSTGRES_HOST_AUTH_METHOD: scram-sha-256
      POSTGRES_INITDB_ARGS: --auth-host=scram-sha-256
  farmos:
    depends_on:
      - db
    image: farmos/farmos:3.3.1
    volumes:
      - '/data/farmos/sites:/opt/drupal/web/sites'
    ports:
      - '3333:80'
    restart: always
```

Happy to provide any other information that may be useful in identifying a cause/resolution.

**Comment by mstenta** [comment]
Hi @followmeongitlab - this issue has come up in the past (eg: https://farmos.discourse.group/t/kml-import-from-google-earth/1931/2).

I think the issue is usually due to the KML that is imported. Sometimes it contains extra or non-standard information which our importer can't handle. You might be able to experiment with editing the KML directly to see if you can figure out what's not working. Any additional information you can figure out would be helpful! I would love to make our code more resilient to the variations of KML files that are out there. It is a surprising non-standard format :sweat_smile:.

Here is what I posted in that forum topic, which is relevant here too:

> We’ve had similar reports over the years that some KML files work and some do not. I think that it depends a lot on what is contained in the KML. I wouldn’t be surprised if removing the “things related to appearance, eg. color and line width” fixed it. But the “outside boundaries” could also have an effect.
> 
> It’s worth noting that the farmOS KML Import module code is custom. We wrote it ourselves as a simple extension of an XML encoder/decoder. It does not use an existing KML library. This means that it’s probably not perfect itself, and can’t handle every possible KML file. We might be able to improve that by using GeoPHP’s KML parsing (which is used when you upload individual KML files to assets/logs and click the “Load from file” button). So that might be a good next step if anyone wants to work on it.
> I think the technical challenge is that we need to be able to intelligently split the file up into multiple distinct assets. Here is the custom KML decoder method in farmOS: [farmOS/modules/core/kml/src/Encoder/Kml.php at 219e00c8744d38159d217c353245006374379e3a · farmOS/farmOS · GitHub](https://github.com/farmOS/farmOS/blob/219e00c8744d38159d217c353245006374379e3a/modules/core/kml/src/Encoder/Kml.php#L54)

**Comment by mstenta** [comment]
> We might be able to improve that by using GeoPHP’s KML parsing (which is used when you upload individual KML files to assets/logs and click the “Load from file” button).

Also worth highlighting: there are two ways of importing KML in farmOS... and they use different code, so you might have better luck with the other one. The difference is that the one you used (`Setup > Import > KML Import`) is for importing MULTIPLE shapes and create multiple assets. But you can also create assets individually, and upload a KML file to them which only contains that asset's geometry. This is more tedious, but worth mentioning in case you are importing single shapes anyway.

More info here: https://farmos.org/guide/import/#geometry

**Comment by followmeongitlab** [comment]
Thanks for the quick response and advice! I apparently skimmed that documentation page rather than reading it the first time. I was confused as to why the upload button was grayed out until I understood I needed to upload it as a file to the asset first before the button would allow me to import. I was able successfully import the parcel following the documentation. Here's to hoping I read the documentation a little closer before opening an issue next time :-) Big fan of the platform so far!
